File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Decorator/Decorator.cpp
#include <iostream>
#include <string>
using namespace std;
class Beverage {
public:
  virtual string getDescription() const = 0;
  virtual double getCost() const = 0;
};

class Coffee : public Beverage{
    public:
        Coffee(){}
        string getDescription() const override{
            return "Coffee";
        };
        double getCost() const override{
            return 1.50;
        };
};
class CondimentDecorator : public Beverage{
    protected:
        Beverage *beverage;
    public:
        CondimentDecorator(Beverage * beverage) : beverage(beverage){}
        virtual string  getDescription() const override{
            return beverage->getDescription()+",";
        };
        virtual double getCost() const override{
            return beverage -> getCost();
        };
};
class Milk : public CondimentDecorator {
public:
  Milk(Beverage* beverage) : CondimentDecorator(beverage) {}
  string getDescription() const override {
    return CondimentDecorator::getDescription() + "Milk";
  };
  double getCost() const override {
    return CondimentDecorator::getCost() + 0.50;
  };
};

class Sugar : public CondimentDecorator {
public:
  Sugar(Beverage* beverage) : CondimentDecorator(beverage) {}
  string getDescription() const override {
    return CondimentDecorator::getDescription() + "Sugar";
  };
  double getCost() const override {
    return CondimentDecorator::getCost() + 0.25;
  };
};

int main() {
  Beverage* beverage = new Coffee();
  cout << beverage->getDescription() << " $" << beverage->getCost() << endl;

  // Add Milk decorator
  beverage = new Milk(beverage);
  cout << beverage->getDescription() << " $" << beverage->getCost() << endl;

  // Add Sugar decorator to the Milk-decorated beverage
  beverage = new Sugar(beverage);
  cout << beverage->getDescription() << " $" << beverage->getCost() << endl;

  return 0;
}

File: 2.OOP/concepts/MediumExample/design_patterns/Structural/Decorator/Decorator.go
package main

import "fmt"

type Beverage interface {
	getDescription() string
	getCost() float64
}
type Coffee struct{}

func (this Coffee) getDescription() string {
	return "Coffee"
}
func (this Coffee) getCost() float64 {
	return 1.5
}

type CondimentDecorator struct {
	beverage Beverage
}

func (this *CondimentDecorator) getDescription() string {
	return this.beverage.getDescription()
}
func (this *CondimentDecorator) getCost() float64 {
	return this.beverage.getCost()
}

type Milk struct {
	CondimentDecorator
}

func (this Milk) getDescription() string {
	return this.CondimentDecorator.getDescription() + ",MILK"
}
func (this Milk) getCost() float64 {
	return this.CondimentDecorator.getCost() + 0.50
}

type Sugar struct {
	CondimentDecorator
}

func (this Sugar) getDescription() string {
	return this.CondimentDecorator.getDescription() + ",Sugar"
}
func (this Sugar) getCost() float64 {
	return this.CondimentDecorator.getCost() + 0.25
}
func main() {
	beverage := Coffee{}
	fmt.Printf("Beverage: %s, Cost: $%.2f\n", beverage.getDescription(), beverage.getCost())

	// Add Milk decorator
	milkBeverage := Milk{CondimentDecorator{beverage}}
	fmt.Printf("Beverage: %s, Cost: $%.2f\n", milkBeverage.getDescription(), milkBeverage.getCost())

	// Add Sugar decorator to the Milk-decorated beverage
	sugarBeverage := Sugar{CondimentDecorator{milkBeverage}}
	fmt.Printf("Beverage: %s, Cost: $%.2f\n", sugarBeverage.getDescription(), sugarBeverage.getCost())
}